File: src/python/asynchrony/02_future_create_task.py
"""
This script demonstrates the use of asyncio.Future when working with async
functions in Python. It shows how to create a future, set a result, and
retrieve the result from a future.

The script contains a synchronous function and an async function, both simulating
a slow computation. We first run the synchronous version and measure the time taken.
Then, we run the asynchronous version with futures and measure the time taken.

This example shows the differences between synchronous and asynchronous approaches
and the effective use of futures to handle results from async functions.
"""

import asyncio
import time


def slow_square_sync(x):
    print(f"Starting slow square computation for {x}")
    time.sleep(2)
    result = x * x
    print(f"Finished slow square computation for {x}")
    return result


async def slow_square_async(x):
    print(f"Starting slow square computation for {x}")
    await asyncio.sleep(2)
    result = x * x
    print(f"Finished slow square computation for {x}")
    return result


def synchronous_execution():
    start_time = time.time()

    result1 = slow_square_sync(3)
    result2 = slow_square_sync(4)

    elapsed_time = time.time() - start_time
    print(f"\nSynchronous execution took {elapsed_time} seconds.")
    print(f"Results: {result1}, {result2}")


async def asynchronous_execution():
    start_time = time.time()

    future1 = asyncio.Future()
    future2 = asyncio.Future()

    task1 = asyncio.create_task(slow_square_async(3))
    task2 = asyncio.create_task(slow_square_async(4))

    future1.set_result(await task1)
    future2.set_result(await task2)

    results = await asyncio.gather(future1, future2)

    elapsed_time = time.time() - start_time
    print(f"\nAsynchronous execution took {elapsed_time} seconds.")
    print(f"Results: {results}")


if __name__ == "__main__":
    synchronous_execution()
    asyncio.run(asynchronous_execution())

File: src/cpp/asynchrony/02_future_create_task.cpp
#include <chrono>
#include <future>
#include <iostream>
#include <thread>

int slow_square_sync(int x) {
  std::cout << "Starting slow square computation for " << x << std::endl;
  std::this_thread::sleep_for(std::chrono::seconds(2));
  int result = x * x;
  std::cout << "Finished slow square computation for " << x << std::endl;
  return result;
}

int slow_square_async(int x) {
  std::cout << "Starting slow square computation for " << x << std::endl;
  std::this_thread::sleep_for(std::chrono::seconds(2));
  int result = x * x;
  std::cout << "Finished slow square computation for " << x << std::endl;
  return result;
}

void synchronous_execution() {
  auto start_time = std::chrono::steady_clock::now();

  int result1 = slow_square_sync(3);
  int result2 = slow_square_sync(4);

  auto elapsed_time = std::chrono::steady_clock::now() - start_time;
  auto elapsed_seconds =
      std::chrono::duration_cast<std::chrono::seconds>(elapsed_time).count();

  std::cout << "\nSynchronous execution took " << elapsed_seconds << " seconds."
            << std::endl;
  std::cout << "Results: " << result1 << ", " << result2 << std::endl;
}

void asynchronous_execution() {
  auto start_time = std::chrono::steady_clock::now();

  auto future1 = std::async(std::launch::async, slow_square_async, 3);
  auto future2 = std::async(std::launch::async, slow_square_async, 4);

  int result1 = future1.get();
  int result2 = future2.get();

  auto elapsed_time = std::chrono::steady_clock::now() - start_time;
  auto elapsed_seconds =
      std::chrono::duration_cast<std::chrono::seconds>(elapsed_time).count();

  std::cout << "\nAsynchronous execution took " << elapsed_seconds
            << " seconds." << std::endl;
  std::cout << "Results: " << result1 << ", " << result2 << std::endl;
}

int main() {
  synchronous_execution();
  asynchronous_execution();

  return 0;
}
